File: src/main/java/com/grandtech/service/UserService.java
package com.grandtech.service;

import com.google.firebase.auth.FirebaseAuthException;
import com.google.firebase.auth.FirebaseToken;
import com.grandtech.model.School;
import com.grandtech.model.Teacher;
import com.grandtech.model.User;
import com.grandtech.utils.ApiResponse;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.transaction.Transactional;

/**
 * Business logic for user account management.
 *
 * <p>Handles token verification, user lookup, and registration of both
 * school and teacher accounts. Profile updates are handled separately.
 */
@ApplicationScoped
public class UserService {

    /** Service used to verify Firebase Authentication ID tokens. */
    @Inject
    FirebaseAuthService firebaseAuthService;

    /**
     * Verifies the supplied Firebase ID token and returns the matching user.
     *
     * @param idToken the raw Firebase ID token from the Authorization header
     * @return the {@link User} whose {@code fedUid} matches the token,
     *         or {@code null} if the account has not been registered yet
     * @throws FirebaseAuthException if the token is invalid, expired, or revoked
     */
    public User getUserFromToken(final String idToken) throws FirebaseAuthException {
        final FirebaseToken token = firebaseAuthService.verifyToken(idToken);
        return User.findByFedUid(token.getUid());
    }

    /**
     * Registers a new teacher account linked to the supplied Firebase token.
     *
     * <p>Creates a row in both the {@code users} and {@code teachers} tables.
     * All profile fields are left empty and must be completed via a
     * subsequent profile-update request.
     *
     * @param idToken the raw Firebase ID token from the Authorization header
     * @return status 200 with the new {@link Teacher} as payload on success,
     *         status 502 with "User already in the system!!!" if the fedUid or email is taken,
     *         or status 500 with "Invalid bearer token!!" if the token is invalid or expired
     */
    @Transactional
    public ApiResponse<Teacher> registerTeacher(final String idToken) {
        final FirebaseToken token;
        try {
            token = firebaseAuthService.verifyToken(idToken);
        } catch (FirebaseAuthException e) {
            return new ApiResponse<>(500, "Invalid bearer token!!", null);
        }
        final String fedUid = token.getUid();
        final String email = token.getEmail();
        if (User.findByFedUid(fedUid) != null
                || Teacher.findByEmail(email) != null
                || School.findByEmail(email) != null) {
            return new ApiResponse<>(502, "User already in the system!!!", null);
        }
        final Teacher teacher = new Teacher();
        teacher.fedUid = fedUid;
        teacher.name = token.getName();
        teacher.email = email;
        teacher.persist();
        return new ApiResponse<>(200, "Success", teacher);
    }

    /**
     * Registers a new school account linked to the supplied Firebase token.
     *
     * <p>Creates a row in both the {@code users} and {@code schools} tables.
     * All profile fields are left empty and must be completed via a
     * subsequent profile-update request.
     *
     * @param idToken the raw Firebase ID token from the Authorization header
     * @return status 200 with the new {@link School} as payload on success,
     *         status 502 with "User already in the system!!!" if the fedUid or email is taken,
     *         or status 500 with "Invalid bearer token!!" if the token is invalid or expired
     */
    @Transactional
    public ApiResponse<School> registerSchool(final String idToken) {
        final FirebaseToken token;
        try {
            token = firebaseAuthService.verifyToken(idToken);
        } catch (FirebaseAuthException e) {
            return new ApiResponse<>(500, "Invalid bearer token!!", null);
        }
        final String fedUid = token.getUid();
        final String email = token.getEmail();
        if (User.findByFedUid(fedUid) != null
                || Teacher.findByEmail(email) != null
                || School.findByEmail(email) != null) {
            return new ApiResponse<>(502, "School already in the system", null);
        }
        final School school = new School();
        school.fedUid = fedUid;
        school.persist();
        return new ApiResponse<>(200, "Success", school);
    }
}

File: src/main/kotlin/com/grandtech/service/UserService.kt
package com.grandtech.service

import com.google.firebase.auth.FirebaseAuthException
import com.grandtech.auth.FirebaseAuthService
import com.grandtech.model.School
import com.grandtech.model.User
import com.grandtech.repository.UserRepository
import com.grandtech.utils.ApiResponse
import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject

/**
 * Business logic for user account management.
 *
 * Handles Firebase token verification, user lookup, and registration of school accounts.
 * All persistence is delegated to [UserRepository].
 */
@ApplicationScoped
class UserService {

    /** Service used to verify Firebase Authentication ID tokens. */
    @Inject
    lateinit var firebaseAuthService: FirebaseAuthService

    /** Repository that executes Neo4j reads and writes for user nodes. */
    @Inject
    lateinit var userRepository: UserRepository

    /**
     * Verifies the supplied Firebase ID token and returns the matching user.
     *
     * @param idToken the raw Firebase ID token from the Authorization header
     * @return the [User] whose `fedUid` matches the token, or null if not registered
     * @throws FirebaseAuthException if the token is invalid, expired, or revoked
     */
    @Throws(FirebaseAuthException::class)
    fun getUserFromToken(idToken: String): User? {
        val token = firebaseAuthService.verifyToken(idToken)
        return userRepository.findByFedUid(token.uid)
    }

    /**
     * Registers a new school account linked to the supplied Firebase token.
     *
     * @param idToken the raw Firebase ID token from the Authorization header
     * @return status 200 with the new [School] on success,
     *         502 if the fedUid or email is already taken,
     *         or 500 if the token is invalid or expired
     */
    fun registerSchool(idToken: String): ApiResponse<School> {
        val token = try {
            firebaseAuthService.verifyToken(idToken)
        } catch (e: FirebaseAuthException) {
            return ApiResponse(500, "Invalid bearer token!!", null)
        }
        val fedUid = token.uid
        val email = token.email
        if (userRepository.existsByFedUid(fedUid) || userRepository.schoolExistsByEmail(email)) {
            return ApiResponse(502, "School already in the system", null)
        }
        val school = School(fedUid = fedUid)
        userRepository.saveSchool(school)
        return ApiResponse(200, "Success", school)
    }
}